### PR TITLE
Auditlog

### PR DESCRIPTION
**Pull request by dffdff2423**
Closes https://github.com/space-wizards/space-station-14/issues/19672
## About the PR

Add an audit log as described in https://github.com/space-wizards/space-station-14/issues/19672. It is currently only viewable inside SS14.Admin because I don't think viewing the audit log would be that useful in game.

Kicks are currently excluded from the log because they are handled by RobustToolbox.
## Why / Balance

The reasons are outlined in the issue. I am kind of surprised there is no way to audit admin actions in the game yet.

## Technical details

Adds a few new tables/migrations to add audit logs to the database. Adds an entry whenever a user takes one of the actions outlined in the issue. I tried to have the logging at the point closest to the database that was possible to reduce code-duplication. The UI for viewing them will be in the SS14.Admin PR that should link to this one below.

## Media

audit-log-example

## Requirements

- [X] I have read and I am following the [Pull Request Guidelines](https://docs.spacestation14.com/en/general-development/codebase-info/pull-request-guidelines.html). I understand that not doing so may get my pr closed at maintainer’s discretion
- [X] I have added screenshots/videos to this PR showcasing its changes ingame, **or** this PR does not require an ingame showcase

## Breaking changes

I don't think there are any.

**Changelog**

:cl: Aquif
ADMIN:
- add: Admin audit logs now exist and can be viewed in SS14.Admin. Audit logs will only exist for *new* actions. Existing actions will not have audit logs.

### CODE REVIEW

**Comment by Geekyhobo** [comment]
This should have a adminflag and permission attached to it, and add that to the ss14admin PR as well

File: Content.Server.Database/remove-migration.sh
Comment (Geekyhobo) [inline Content.Server.Database/remove-migration.sh:1]: is this needed

Comment (dffdff2423) [reply Content.Server.Database/remove-migration.sh:1]: It is not really needed, but I found it useful to have a counterpart to the add-migration.sh script when I was debugging migrations. I can remove it if you think it is unnecessary.

File: Content.Shared.Database/AuditLogType.cs [hunk 1]
```diff
@@ -0,0 +1,16 @@
...
+public enum AuditLogType : uint
+{
+    Unknown = 0, // Do not use this is a fallback
+    Whitelist = 1,
+    BanExemption = 2,
+    Note = 3,
+    Message = 4,
+    Watchlist = 5,
+    RoleBan = 6,
+    SeverBan = 7,
```
Comment (metalgearsloth) [inline Content.Shared.Database/AuditLogType.cs:13]: Server

**Comment by dffdff2423** [comment]
Branch got a little messed up because the submodule script does not work when your ss14 repo is a submodule and I forgot I did not have it installed. I should have fixed it now.

**Review by PJB3005 — CHANGES_REQUESTED** [review]

File: Content.Server.Database/Model.cs [hunk 2]
```diff
@@ -628,6 +654,41 @@ public class Server
         public List<ConnectionLog> ConnectionLogs { get; set; } = default!;
     }
 
+    [PrimaryKey(nameof(Id))]
+    public class AuditLog
```
[the remark this answers is not in the document]
Comment (dffdff2423) [reply Content.Server.Database/Model.cs:658]: I remember there being *some* reason why I did not do that, but this was so long ago I can not remember. I will take another look at it.

File: Content.Server/Administration/Commands/BanExemptionCommands.cs [hunk 3]
```diff
@@ -22,6 +27,13 @@ public override async void Execute(IConsoleShell shell, string argStr, string[]
             return;
         }
 
+        if (shell.Player == null)
+        {
+            shell.WriteError(Loc.GetString("shell-uid-failure"));
+            return;
+        }
```
[the remark this answers is not in the document]
Comment (dffdff2423) [reply Content.Server/Administration/Commands/BanExemptionCommands.cs:34]: I have no idea why I thought this was acceptable.

**Comment by dffdff2423** [comment]
I will try to get around to re-vising this some time this weekend, my free-time is significantly lower right now for a few IRL reasons but should clear up soon.

**Comment by dffdff2423** [comment]
Discussed on discord with CE here https://discord.com/channels/310555209753690112/770682801607278632/1295911799220342867

**Comment by dffdff2423** [comment]
"Recent events" and the fact that I have multiple small and easy to merge PRs that have been open for two and a half months without a review have killed my motivation to make major contributions to wizden. I am going to close this PR so it is not hanging over me while I spend my limited free time working on other projects. I may return to this in a few months, or I may not. Please mark it as derelict. Thank you for understanding. 

Anybody who wishes to pick up this PR and make it into a state that is mergeable, feel free to contact me here or on discord dffdff2423 (551544963201105931) if you have questions. I would strongly recommend reading the above discussion I had with CE.

**Pull request closed without merging**